pkg/adaptor/mongodb/client.go

```diff
@@ -0,0 +1,257 @@
+package mongodb
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	// DefaultURI is the default endpoint of MongoDB on the local machine.
+	// Primarily used when initializing a new Client without a specific URI.
+	DefaultURI = "mongodb://127.0.0.1:27017/test"
+
+	// DefaultSessionTimeout is the default timeout after which the
+	// session times out when unable to connect to the provided URI.
+	DefaultSessionTimeout = 10 * time.Second
+)
+
+var (
+	// DefaultSafety is the default saftey mode used for the underlying session.
+	// These default settings are only good for local use as it makes not guarantees for writes.
+	DefaultSafety = mgo.Safe{}
+
+	_ client.Client = &Client{}
+)
+
+// InvalidURIError wraps the underlying error when the provided URI is not parsable by mgo
+type InvalidURIError struct {
+	uri      string
+	mgoError string
+}
+
+func (e InvalidURIError) Error() string {
+	return fmt.Sprintf("Invalid URI (%s), %s", e.uri, e.mgoError)
+}
+
+// InvalidTimeoutError wraps the underlying error when the provided is not parsable time.ParseDuration
+type InvalidTimeoutError struct {
+	timeout string
+}
+
+func (e InvalidTimeoutError) Error() string {
+	return fmt.Sprintf("Invalid Timeout, %s", e.timeout)
+}
+
+// ClientOptionFunc is a function that configures a Client.
+// It is used in NewClient.
+type ClientOptionFunc func(*Client) error
+
+// Client represents a client to the underlying MongoDB source.
+type Client struct {
+	uri string
+
+	safety         mgo.Safe
+	tlsConfig      *tls.Config
+	sessionTimeout time.Duration
+	tail           bool
+
+	mgoSession *mgo.Session
+	bulkWriter chan message.Msg
+}
+
+// NewClient creates a new client to work with MongoDB.
+//
+// The caller can configure the new client by passing configuration options
+// to the func.
+//
+// Example:
+//
+//   client, err := NewClient(
+//     SetURI("mongodb://localhost:27017"),
+//     SetTimeout("30s"))
+//
+// If no URI is configured, it uses defaultURI by default.
+//
+// An error is also returned when some configuration option is invalid
+func NewClient(options ...ClientOptionFunc) (*Client, error) {
+	// Set up the client
+	c := &Client{
+		uri:            DefaultURI,
+		sessionTimeout: DefaultSessionTimeout,
+		safety:         DefaultSafety,
+		tlsConfig:      nil,
+		tail:           false,
+		bulkWriter:     nil,
+	}
+
+	// Run the options on it
+	for _, option := range options {
+		if err := option(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// SetURI defines the full connection string of the MongoDB database.
```

Comment:
gotcha, I was mostly mimicking https://github.com/olivere/elastic/blob/v5.0.13/client.go so didn't have a whole lot of other libraries for reference